From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Wed, 18 Feb 2026 00:15:08 +0800
Subject: [PATCH] Leaves: Fakeplayer


diff --git a/src/main/java/me/earthme/luminol/utils/FoliaServerWaypointManager.java b/src/main/java/me/earthme/luminol/utils/FoliaServerWaypointManager.java
index 389a31b14912a2dfc8e8b20c104505d211d7f8ad..04b9e3a432181c9698e1a4c147b420c230b66a36 100644
--- a/src/main/java/me/earthme/luminol/utils/FoliaServerWaypointManager.java
+++ b/src/main/java/me/earthme/luminol/utils/FoliaServerWaypointManager.java
@@ -55,6 +55,12 @@ public class FoliaServerWaypointManager extends ServerWaypointManager {
             return;
         }
 
+        // Leaves start - fakeplayer
+        if (waypoint instanceof org.leavesmc.leaves.bot.ServerBot bot && !bot.getConfigValue(org.leavesmc.leaves.bot.agent.Configs.ENABLE_LOCATOR_BAR)) {
+            return;
+        }
+        // Leaves end - fakeplayer
+
         this.waypoints.add(waypoint);
 
         for (ServerPlayer toCreateFor : this.trackingPlayers) {
@@ -111,6 +117,12 @@ public class FoliaServerWaypointManager extends ServerWaypointManager {
         scheduleIfOffTarget(player, () -> {
             this.trackingPlayers.add(player);
 
+            // Leaves start - fakeplayer
+            if (player instanceof org.leavesmc.leaves.bot.ServerBot) {
+                return;
+            }
+            // Leaves end - fakeplayer
+
             for (WaypointTransmitter transmitter : this.waypoints) {
                 this.createConnection(player, transmitter);
             }
diff --git a/src/main/java/org/leavesmc/leaves/plugin/ServerFeatureManager.java b/src/main/java/org/leavesmc/leaves/plugin/ServerFeatureManager.java
index 063cc27df9c8f3ad5c8db5cedcaa3d2bccf6dc24..5ad6aab5e72e27bb402494873770ab5822601939 100644
--- a/src/main/java/org/leavesmc/leaves/plugin/ServerFeatureManager.java
+++ b/src/main/java/org/leavesmc/leaves/plugin/ServerFeatureManager.java
@@ -11,14 +11,7 @@ public class ServerFeatureManager implements FeatureManager {
     private final Set<String> availableFeatures = new HashSet<>();
 
     private ServerFeatureManager() {
-        if (false) { // Luminol not provide this
-            availableFeatures.addAll(Set.of(
-                    FAKEPLAYER,
-                    PHOTOGRAPHER,
-                    BYTEBUF,
-                    UPDATE_SUPPRESSION_EVENT
-            ));
-        }
+        availableFeatures.add(FAKEPLAYER); // Lophine - fakeplayer support
         if (Boolean.getBoolean("leavesclip.enable.mixin") || Boolean.getBoolean("hyacinthusclip.enable.mixin")) {
             availableFeatures.add(MIXIN);
         }
