From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 22 Mar 2026 01:39:17 +0800
Subject: [PATCH] Carpet features


diff --git a/src/main/java/me/earthme/luminol/config/ConfigManager.java b/src/main/java/me/earthme/luminol/config/ConfigManager.java
index 8d6b1e2ff0193b69636a52485c7fdaa4b7d6ff15..697cc65e8487fe78a4f8ae0435c22c9ef8cdc80c 100644
--- a/src/main/java/me/earthme/luminol/config/ConfigManager.java
+++ b/src/main/java/me/earthme/luminol/config/ConfigManager.java
@@ -25,6 +25,13 @@ public class ConfigManager {
     public static void initConfigs() {
         configfiles.put("luminol", ConfigsInstance.of("luminol", "me.earthme.luminol.config.modules"));
         configfiles.put("lophine", ConfigsInstance.of("lophine", "fun.bm.lophine.config.modules")); // add lophine global config
+        configfiles.put("lophine_carpet", ConfigsInstance.of(
+                new java.io.File("lophine_config"),
+                "lophine_carpet",
+                "lophine_carpet_config.toml",
+                "lophinecarpetconfig",
+                "fun.bm.lophine.carpet.config.modules"
+        )); // add lophine carpet config
         preLoad();
     }
 
