From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 22 Mar 2026 01:39:17 +0800
Subject: [PATCH] Carpet features


diff --git a/src/main/java/me/earthme/luminol/config/ConfigManager.java b/src/main/java/me/earthme/luminol/config/ConfigManager.java
index 05e1aa84586284ac0389e7023f975bfb08438535..5a8c94337e42584ddaf25ef289327777e7ac675e 100644
--- a/src/main/java/me/earthme/luminol/config/ConfigManager.java
+++ b/src/main/java/me/earthme/luminol/config/ConfigManager.java
@@ -25,6 +25,15 @@ public class ConfigManager {
     public static void initConfigs() {
         registerConfig("luminol", ConfigsInstance.of("luminol", "me.earthme.luminol.config.modules"));
         registerConfig("lophine", ConfigsInstance.of("lophine", "fun.bm.lophine.config.modules")); // add lophine global config
+        // Lophine start - add carpet config
+        registerConfig("lophine_carpet", ConfigsInstance.of(
+                new java.io.File("lophine_config"),
+                "lophine_carpet",
+                "lophine_carpet_config.toml",
+                "lophinecarpetconfig",
+                "fun.bm.lophine.carpet.config.modules"
+        ));
+        // Lophine end - add carpet config
         preLoad();
     }
 
