From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 29 Sep 2025 22:56:03 +0800
Subject: [PATCH] Add config to enable scoreboard command


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index 6d27350cdb2e699fcc541bc96a5af06a09e0dfd5..01784d3feaa0dce02f0c26b9a635877b735e6abf 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -244,7 +244,11 @@ public class Commands {
         RotateCommand.register(this.dispatcher);
         SayCommand.register(this.dispatcher);
         //ScheduleCommand.register(this.dispatcher); // Folia - region threading
-        //ScoreboardCommand.register(this.dispatcher, context); // Folia - region threading
+        // Lophine start - Add a config to enable scoreboard command
+        if (fun.bm.lophine.config.modules.experiment.CommandConfig.scoreboard) {
+            ScoreboardCommand.register(this.dispatcher, context); // Folia - region threading
+        }
+        // Lophine start - Add a config to enable scoreboard command
         SeedCommand.register(this.dispatcher, commandSelection != Commands.CommandSelection.INTEGRATED);
         VersionCommand.register(this.dispatcher, commandSelection != Commands.CommandSelection.INTEGRATED);
         SetBlockCommand.register(this.dispatcher, context);
