From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 17 Nov 2025 18:41:08 +0800
Subject: [PATCH] Instant Block Updater


diff --git a/io/papermc/paper/threadedregions/RegionizedWorldData.java b/io/papermc/paper/threadedregions/RegionizedWorldData.java
index 8d6ee598bdbbf9e238328abe0646234af26914d4..6d3bbfcc736ea357084c4668d1a831374edf5f10 100644
--- a/io/papermc/paper/threadedregions/RegionizedWorldData.java
+++ b/io/papermc/paper/threadedregions/RegionizedWorldData.java
@@ -506,7 +506,7 @@ public final class RegionizedWorldData {
     public long lastMidTickExecuteFailure;
     // From Level
     public boolean populating;
-    public final CollectingNeighborUpdater neighborUpdater;
+    public final net.minecraft.world.level.redstone.NeighborUpdater neighborUpdater; // Lophine - Instant Block Updater
     public boolean captureBlockStates = false;
     public boolean captureTreeGeneration = false;
     public final Map<BlockPos, CraftBlockState> capturedBlockStates = new java.util.LinkedHashMap<>(); // Paper
@@ -567,7 +567,11 @@ public final class RegionizedWorldData {
         this.world = world;
         this.blockLevelTicks = new LevelTicks<>(world::isPositionTickingWithEntitiesLoaded, world, true);
         this.fluidLevelTicks = new LevelTicks<>(world::isPositionTickingWithEntitiesLoaded, world, false);
-        this.neighborUpdater = new CollectingNeighborUpdater(world, world.neighbourUpdateMax);
+        // Lophine start - Instant Block Updater
+        this.neighborUpdater = fun.bm.lophine.carpet.config.modules.GeneralCompatConfig.instantBlockUpdaterReintroduced
+            ? new net.minecraft.world.level.redstone.InstantNeighborUpdater(world)
+            : new CollectingNeighborUpdater(world, world.neighbourUpdateMax);
+        // Lophine end - Instant Block Updater
         this.nearbyPlayers = new NearbyPlayers(world);
         this.wireHandler = new alternate.current.wire.WireHandler(world);
         this.turbo = new io.papermc.paper.redstone.RedstoneWireTurbo((RedStoneWireBlock)Blocks.REDSTONE_WIRE);
@@ -896,4 +900,4 @@ public final class RegionizedWorldData {
     public int getChunkCount() {
         return this.chunks.size();
     }
-}
\ No newline at end of file
+}
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 3e7420cd98504732db17cb6b0fcfdfbb8deeb552..257b440a0eea379433e9a249175c8eb16c37accb 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1001,11 +1001,15 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         // Paper - rewrite chunk system
         profiler.pop();
         profiler.push("debugSynchronizers");
-        if (this.debugSynchronizers.hasAnySubscriberFor(DebugSubscriptions.NEIGHBOR_UPDATES)) {
-            regionizedWorldData.neighborUpdater // Folia - region threading
-                .setDebugListener(blockPos -> this.debugSynchronizers.broadcastEventToTracking(blockPos, DebugSubscriptions.NEIGHBOR_UPDATES, blockPos));
-        } else {
-            regionizedWorldData.neighborUpdater.setDebugListener(null); // Folia - region threading
+        // Lophine start - Instant Block Updater
+        if (regionizedWorldData.neighborUpdater instanceof net.minecraft.world.level.redstone.CollectingNeighborUpdater neighborUpdater) {
+            if (this.debugSynchronizers.hasAnySubscriberFor(DebugSubscriptions.NEIGHBOR_UPDATES)) {
+                neighborUpdater // Folia - region threading
+                    .setDebugListener(blockPos -> this.debugSynchronizers.broadcastEventToTracking(blockPos, DebugSubscriptions.NEIGHBOR_UPDATES, blockPos));
+            } else {
+                neighborUpdater.setDebugListener(null); // Folia - region threading
+            }
+        // Lophine end - Instant Block Updater
         }
 
         this.debugSynchronizers.tick(this.server.debugSubscribers());
